### PR TITLE
Inference custom model example

### PR DESCRIPTION
**Pull request by RR4787**

### CODE REVIEW

File: examples/inference/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Inference with Custom Models on the MosaicML platform
+
+[MosaicML Inference]() enables easy deployment of machine learning models. In this folder, we provide an example for how to create and deploy your own model on the MosaicML platform.
+
+You'll find in this folder:
+
+- `hf_instructor_model_class.py` - a custom model class that implements the Hugging Face large instructor
```
Comment (dblalock) [inline examples/inference/README.md:7]:
```suggestion
- `hf_instructor_model_class.py` - a custom model class that implements the Hugging Face model [instructor-large](https://huggingface.co/hkunlp/instructor-large)
```

File: examples/inference/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+- `custom_mosaic_gpt.yaml` - a script to deploy the model on the MosaicML platform
+
+## Prerequisites
+
+The only thing you need to deploy this model is access to MosaicML's inference service. [Contact us to get started]().
```
Comment (dblalock) [inline examples/inference/README.md:12]:
```suggestion
The only thing you need to deploy this model is access to MosaicML's inference service. [Contact us to get started](https://forms.mosaicml.com/demo).
```
(Note that this URL might change in like a day)

File: examples/inference/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+## Configuring the Model
+
+Currently, we offer support for any hugging face model and custom models that follow a model class spec. The provided `.yaml` file is configured to deploy the custom instructor model class. Information on the configuration options can be found [here]().
```
Comment (dblalock) [inline examples/inference/README.md:16]:
```suggestion
Currently, we offer support for any Hugging Face model and custom models that implement a particular interface. The provided `.yaml` file is configured to deploy the custom instructor model class. Information on the configuration options can be found [here]().
```
Can we document this interface here? Or is there a class they have to subclass or something? Mostly just want to link to something specific enough for people to use it.

File: examples/inference/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+# Deploying the Model
+
+Deploying the model is as simple as running `mcli deploy -f yamls/instructor_model.yanl`.
```
Comment (dblalock) [inline examples/inference/README.md:20]:
```suggestion
Deploying the model is as simple as running `mcli deploy -f yamls/instructor_model.yaml`.
```

File: examples/inference/src/hf_instructor_model_class.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+    def setup(self):
+        self.model = INSTRUCTOR(self.MODEL_NAME)
+
+    def predict(self, **inputs: Dict[str, Any]):
+        # input_strings in format of [[<instruction>, <sentence>]]
+        if "input_strings" not in inputs:
+            raise KeyError("input_strings key not in inputs")
+        
+        print(inputs)
```
Comment (dblalock) [inline examples/inference/src/hf_instructor_model_class.py:33]:
```suggestion
```

**Review by dblalock — APPROVED** [review]
I'm approving this so we have at least a placeholder to link to in the blog post, but we need at least a few changes in a follow-up PR:

- a requirements.txt. In particular, `from InstructorEmbedding import INSTRUCTOR` doesn't look like it should work.
- some sort of unit testing in a tests/ subdirectory. Doesn't have to be fancy--just something that catches bitrot.
- The addition of this subdirectory to the list of subdirs we lint + test
- Simple docstrings in the Python classes

Also, I got an error when trying to run the example command, so maybe we have missing installation steps or something? Hopefully a link to the inference docs will be enough.
```
~/code/examples/examples/inference roshan/inference_hf_model *1 ?1 ❯ mcli deploy -f yamls/instructor_model.yaml
✗  Error 401: Not Authorized       
```

**Review by hanlint — CHANGES_REQUESTED** [review]
Please open this PR into: https://github.com/mosaicml/llm-foundry

File: examples/inference/README.md (hunk 3 above)
Comment (RR4787) [reply examples/inference/README.md]: Its a class they have to inherit from. I've added a link to the mcli inference docs where the class is listed

**Pull request merged**